c10/core/impl/cow/state.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <c10/core/impl/cow/simulator.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/Optional.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <cstdint>
+#include <mutex>
+
+namespace c10::impl::cow {
+
+// Responsible for managing the copy-on-write simulation state
+// machine.
+class C10_API State {
+ public:
+  // Gets the current storage generation.
+  auto storage_generation() -> std::uint64_t;
+
+  // Bumps the generation if simulator is non-null. If non-null,
+  // simulator must be the result of a previous call to
+  // simulate_copy_on_write on this instance.
+  auto maybe_bump(cow::Simulator* maybe_simulator) -> void;
+
+  // Simulates a copy on write.
+  //
+  // maybe_simulator comes from the tensor that would be receiving a
+  // copy on write. If non-null, then the result gets the generation
+  // number from this value.
```

Comment:
The lifecycle of default_simulator from null to non-null should be clearly described, it's not easy to figure out what this is doing here